nengo/tests/test_copy.py

```diff
@@ -0,0 +1,228 @@
+from copy import copy
+import warnings
+
+import numpy as np
+import pytest
+
+import nengo
+from nengo.exceptions import NetworkContextError, NotAddedToNetworkWarning
+from nengo.params import params
+from nengo.utils.compat import is_array_like, pickle
+from nengo.utils.testing import warns
+
+
+@pytest.fixture(scope='module', autouse=True)
+def error_on_not_added_to_network_warning(request):
+    catcher = warnings.catch_warnings(record=True)
+    catcher.__enter__()
+    warnings.simplefilter(
+        'error', category=NotAddedToNetworkWarning)
+    request.addfinalizer(lambda: catcher.__exit__(None, None, None))
+
+
+def assert_is_copy(cp, original):
+    assert cp is not original  # ensures separate parameters
+    for param in params(cp):
+        param_inst = getattr(cp, param)
+        if isinstance(param_inst, nengo.solvers.Solver) or isinstance(
+                param_inst, nengo.base.NengoObject):
+            assert param_inst is getattr(original, param)
+        elif is_array_like(param_inst):
+            assert np.all(param_inst == getattr(original, param))
+        else:
+            assert param_inst == getattr(original, param)
+
+
+def assert_is_deepcopy(cp, original):
+    assert cp is not original  # ensures separate parameters
+    for param in params(cp):
+        param_inst = getattr(cp, param)
+        if isinstance(param_inst, nengo.solvers.Solver) or isinstance(
+                param_inst, nengo.base.NengoObject):
+            assert_is_copy(param_inst, getattr(original, param))
+        elif is_array_like(param_inst):
+            assert np.all(param_inst == getattr(original, param))
+        else:
+            assert param_inst == getattr(original, param)
+
+
+class CopyTest(object):
```

Comment:
I think I'm asking for a silly comment again, but I'd just like to put this out there. A comment indicating that this is a base class that is inherited for testing copying specific object might be helpful? I just spend a couple of minutes trying to figure out how this works, but are we just assuming that whoever is reading these tests is better at PyTest than I am?